Libraries/LibGUI/FilteringProxyModel.cpp

```diff
@@ -0,0 +1,126 @@
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <LibGUI/FilteringProxyModel.h>
+#include <LibGUI/TextBox.h>
+
+namespace GUI {
+
+FilteringProxyModel::FilteringProxyModel(NonnullRefPtr<TextBox> textbox, NonnullRefPtr<Model> model)
+    : m_model(move(model))
+    , m_textbox(move(textbox))
+{
+    m_textbox->on_change = [&] {
+        if (m_is_searching) {
+            m_search_term = m_textbox->text();
+            search();
+        }
+    };
+}
+
+ModelIndex FilteringProxyModel::index(int row, int column, const ModelIndex& parent_index) const
+{
+    return m_model->index(row, column, parent_index);
+}
+
+int FilteringProxyModel::row_count(const ModelIndex&) const
+{
+    return m_matching_indices.size();
+}
+
+int FilteringProxyModel::column_count(const ModelIndex& index) const
+{
+    if (!index.is_valid())
+        return {};
+
+    if ((size_t)index.row() > m_matching_indices.size() || index.row() < 0)
+        return 0;
+
+    return m_model->column_count(m_matching_indices[index.row()]);
+}
+
+Variant FilteringProxyModel::data(const ModelIndex& index, Role role) const
+{
+    if (!index.is_valid())
+        return {};
+
+    if ((size_t)index.row() > m_matching_indices.size() || index.row() < 0)
+        return 0;
+
+    return m_model->data(m_matching_indices[index.row()], role);
+}
+
+void FilteringProxyModel::update()
+{
+    m_model->update();
+    did_update();
+}
+
+void FilteringProxyModel::search()
+{
+    update();
+    m_matching_indices.clear();
+    SearchData search_data { m_search_term, m_full_search };
+
+    Function<void(ModelIndex&)> add_matching = [&](ModelIndex& parent_index) {
+        for (auto i = 0; i < m_model->row_count(parent_index); ++i) {
+            auto index = m_model->index(i, 0, parent_index);
+            if (!index.is_valid())
+                continue;
+
+            index.set_role_specific_data(&search_data);
```

Comment:
I, uh, don't think I understand or like this. Mutating indexes after they've been created, from the outside?..

Reply:
Yeah, hold on to that thought, messing with `Model::data` is gonna cost us a few hundred years of copy-pastes

Reply:
So this role specific data is basically an additional argument to `data()`, but done this way to avoid changing the million declarations?

Reply:
Yep.
Now changing the million declarations.